lightning/src/util/ser_macros.rs

```diff
@@ -1,3 +1,91 @@
+macro_rules! encode_tlv {
+	($stream: expr, {$(($type: expr, $field: expr)),*}) => { {
+		use bitcoin::consensus::Encodable;
+		use bitcoin::consensus::encode::{Error, VarInt};
+		use util::ser::{WriterWriteAdaptor, LengthCalculatingWriter};
+		$(
+			VarInt($type).consensus_encode(WriterWriteAdaptor($stream))
+				.map_err(|e| if let Error::Io(ioe) = e { ioe } else { unreachable!() })?;
+			let mut len_calc = LengthCalculatingWriter(0);
+			$field.write(&mut len_calc)?;
+			VarInt(len_calc.0 as u64).consensus_encode(WriterWriteAdaptor($stream))
+				.map_err(|e| if let Error::Io(ioe) = e { ioe } else { unreachable!() })?;
+			$field.write($stream)?;
+		)*
+	} }
+}
+
+macro_rules! encode_varint_length_prefixed_tlv {
+	($stream: expr, {$(($type: expr, $field: expr)),*}) => { {
+		use bitcoin::consensus::Encodable;
+		use bitcoin::consensus::encode::{Error, VarInt};
+		use util::ser::{WriterWriteAdaptor, LengthCalculatingWriter};
+		let mut len = LengthCalculatingWriter(0);
+		encode_tlv!(&mut len, {
+			$(($type, $field)),*
+		});
+		VarInt(len.0 as u64).consensus_encode(WriterWriteAdaptor($stream))
+			.map_err(|e| if let Error::Io(ioe) = e { ioe } else { unreachable!() })?;
+		encode_tlv!($stream, {
+			$(($type, $field)),*
+		});
```

Comment:
Right, I believe that is correct. Hopefully LLVM will be smart enough to optimize out the template writes here into something sensible, though I haven't gone and checked what actually gets generated.

I'm not sure how to do it much better, sadly, however, given the length tags are variable-length (have I mentioned how much I think variable length tags are a great way to make sure everything is maximally inefficient and insecure?) you're forced to at minimum do the whole thing twice, at without building a cache of lengths do it three times.

I did go ahead and write out the length calculation in the leng_prefixed macro so that its only 3, which I guess is at least a bit better?

Reply:
Totally feel your sentiment regarding variable-length tags. However, can't each field be encoded/serialized only once by keeping the serialized data in memory before writing it to the stream? That way the length could be computed and written first.

I guess that's why I was asking about tradeoffs.

Reply:
Right, yea, we could do that, but I'd really rather not add extra allocations in our serialization pipeline. If LLVM optimizes away most of the calls as-is, then extra allocations would be performance suicide, and if it doesn't, hitting the heap is probably still worse than the alternative.